src/node/interfaces.cpp

```diff
@@ -659,9 +717,96 @@ class ChainImpl : public Chain
+            if (!active_chain.Contains(block_to_test)) {
+                // if the bestblock is not part of the mainchain, find the fork
+                // and make sure we have all data down to the fork
+                block_to_test = active_chain.FindFork(block_to_test);
+            }
+            const CBlockIndex* block = active_chain.Tip();
+            prune_violation = true;
+            // check backwards from the tip if we have all block data until we reach the indexes bestblock
+            while (block_to_test && block && (block->nStatus & BLOCK_HAVE_DATA)) {
+                if (block_to_test == block) {
+                    prune_violation = false;
+                    break;
+                }
+                // block->pprev must exist at this point, since block_to_test is part of the chain
+                // and thus must be encountered when going backwards from the tip
+                assert(block->pprev);
+                block = block->pprev;
+            }
+        }
+        return !prune_violation;
+    }
+    std::unique_ptr<Handler> attachChain(std::shared_ptr<Notifications> notifications, const CBlockLocator& locator, const NotifyOptions& options) override
+    {
+        std::unique_ptr<NotificationsHandlerImpl> handler;
+        std::optional<interfaces::BlockInfo> block_info;
+        // std::promise to wait for the first blockConnected notification, so
+        // caller is notified about the starting block before this returns.
+        // It might be good in the future to return without waiting for this,
+        // and allow more initialization code to run in parallel.
+        std::promise<void> promise;
+        {
+        // Lock cs_main while finding forks and determining which blocks to
+        // rescan. Release cs_main while processing blocks and while calling
+        // RegisterSharedValidationInterface in the notifications thread.
+        //
+        // To prevent older, stale notifications currently in the validation
+        // queue from being received, it is important to not to register for
+        // notifications immediately, but to delay registration with
+        // CallFunctionInValidationInterfaceQueue, registering the notification
+        // proxy inside the queue, so older notifications in the queue before it
+        // get processed before the proxy is registered.
+        //
+        // To prevent new notifications that may be happening in the background
+        // from being lost, it is important to keep cs_main locked while calling
+        // CallFunctionInValidationInterfaceQueue, so the new notifications will
+        // be added to the queue after the queue entry which connects the proxy.
+        AssertLockNotHeld(::cs_main);
+        LOCK(cs_main);
+        const CChainState& active = Assert(m_node.chainman)->ActiveChainstate();
+        const CBlockIndex* locator_block = locator.IsNull() ? nullptr : active.FindForkInGlobalIndex(locator);
```

Comment:
One general point (if it's too unrelated to this PR, we could also discuss it elsewhere):

I think that the current way of initialising indexes by calling `FindForkInGlobalIndex()` for the saved locator and setting this as the best block is not compatible with indexes that have state beyond the best block, such as coinstatsindex:
For this index, we need to call `CustomRemove()` (or `Rewind` on master) to revert the actions of each block and adjust the muhash. This is not done during Init, so the coinstatsindex currently cannot deal with the situation where the saved best block of the index is not part of the best chain.

One solution for this would be not to call `FindForkInGlobalIndex` but instead read in the first block stored in the locator and then revert blocks step by step up to the fork point - this is necessary for coinstatsindex, but it adds the additional requirement that all blocks between the forkpoint and the saved block must be available, which reduces portability of index data and is something that e.g. txindex doesn't really need.  

I did a rough implementation for fixing the coinstatsindex init sequence in [this branch](https://github.com/mzumsande/bitcoin/tree/202205_index_init_rewind), which I think works but (at least on first glance) seems to be really incompatible with the approach here, so I wouldn't just PR it.

What do you think about this problem?

Reply:
> I think that the current way of initialising indexes by calling FindForkInGlobalIndex() for the saved locator and setting this as the best block

You're right but this is easy to fix by just replacing `FindForkInGlobalIndex()` call with a `LookupBlockIndex` call. There should be no need to add additional remove/rewind calls, though. Existing code should be aware of forks and how to handle them.

Reply:
Great, thanks! I missed that `NextSyncBlock()` goes back to the fork point already in master during sync. If it's that easy to fix, I might open a PR vs master. Here, this can be resolved.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/24230#discussion_r874801114

> Great, thanks! I missed that `NextSyncBlock()` goes back to the fork point already in master during sync. If it's that easy to fix, I might open a PR vs master. Here, this can be resolved.

Yes, this is a good catch and test on your branch looks good, but the fix could be simpler and more generic.

Just to close the thread, the issue described here is pre-existing, and basically orthogonal to this PR even if it conflicts a little. The issue just happens if there is a reorg away from the last block that was indexed while the index is shut down.